pkg/assembler/backends/testing/src.go

```diff
@@ -17,120 +17,285 @@ package testing
+// The optional filter allows restricting output (on selection operations).
+func buildSourceResponse(id nodeID, filter *model.SourceSpec) (*model.Source, error) {
+	if filter != nil && filter.ID != nil {
+		filteredID, err := strconv.Atoi(*filter.ID)
+		if err != nil {
+			return nil, err
+		}
+		if nodeID(filteredID) != id {
+			return nil, nil
 		}
 	}
-	return n
-}
 
-func checkQualifier(n *model.SourceName, qualifier string) bool {
-	if qualifier != "" {
-		pair := strings.Split(qualifier, "=")
-		if pair[0] == "tag" {
-			if n.Tag != nil {
-				return *n.Tag == pair[1]
-			}
-		} else {
-			if n.Commit != nil {
-				return *n.Commit == pair[1]
-			}
-		}
+	node, ok := index[id]
+	if !ok {
+		return nil, gqlerror.Errorf("ID does not match existing node")
 	}
-	return false
-}
 
-// Query Source
+	snl := []*model.SourceName{}
+	if nameNode, ok := node.(*srcNameNode); ok {
+		sn := model.SourceName{
+			// IDs are generated as string even though we ask for integers
+			// See https://github.com/99designs/gqlgen/issues/2561
+			ID:   fmt.Sprintf("%d", nameNode.id),
+			Name: nameNode.name,
+		}
+		sn.Tag = &nameNode.tag
 
-func (c *demoClient) Sources(ctx context.Context, sourceSpec *model.SourceSpec) ([]*model.Source, error) {
-	var sources []*model.Source
-	for _, s := range c.sources {
-		if sourceSpec.Type == nil || s.Type == *sourceSpec.Type {
-			newSource, err := filterSourceNamespace(s, sourceSpec)
-			if err != nil {
-				return nil, err
-			}
-			if newSource != nil {
-				sources = append(sources, newSource)
-			}
+		sn.Commit = &nameNode.commit
+
+		if filter != nil && noMatch(filter.Name, sn.Name) {
+			return nil, nil
+		}
+		if filter != nil && noMatchPtr(filter.Tag, sn.Tag) {
+			return nil, nil
+		}
+		if filter != nil && noMatchPtr(filter.Commit, sn.Commit) {
+			return nil, nil
+		}
+		snl = append(snl, &sn)
+		node = index[nameNode.parent]
+	}
+
+	snsl := []*model.SourceNamespace{}
+	if nameStruct, ok := node.(*srcNameStruct); ok {
+		sns := model.SourceNamespace{
+			// IDs are generated as string even though we ask for integers
+			// See https://github.com/99designs/gqlgen/issues/2561
+			ID:        fmt.Sprintf("%d", nameStruct.id),
+			Namespace: nameStruct.namespace,
+			Names:     snl,
+		}
+		if filter != nil && noMatch(filter.Namespace, sns.Namespace) {
+			return nil, nil
 		}
+		snsl = append(snsl, &sns)
+		node = index[nameStruct.parent]
 	}
-	return sources, nil
+
+	namespaceStruct, ok := node.(*srcNamespaceStruct)
+	if !ok {
+		return nil, gqlerror.Errorf("ID does not match expected node type")
+	}
+	s := model.Source{
+		// IDs are generated as string even though we ask for integers
+		// See https://github.com/99designs/gqlgen/issues/2561
+		ID:         fmt.Sprintf("%d", namespaceStruct.id),
+		Type:       namespaceStruct.typeKey,
+		Namespaces: snsl,
+	}
+	if filter != nil && noMatch(filter.Type, s.Type) {
+		return nil, nil
+	}
+	return &s, nil
 }
 
+// Query Source
```

Comment:
This should be on top of `Sources`?

And the `filterSourceNamespace` should no longer be in use?

Reply:
keeping `filterSourceNamespace` for now as others depend on it. Will remove as part of the cleanup